### PR TITLE
normalizeForVisualization() and other methods applying on a ProjectedCRS: do not mess the derivingConversion object of the original object (fixes #1736)

### LINKED ISSUE #1736: proj_create_operations intermittent failures
I've struggled to narrow down the situation which triggers this, but unfortunately this is the best I can get....

Given the following code:

```
  PJ_CONTEXT *c = ...; // something, see below!
  PJ_OPERATION_FACTORY_CONTEXT *operationContext = proj_create_operation_factory_context( c, nullptr );
  proj_operation_factory_context_set_grid_availability_use( c, operationContext, PROJ_GRID_AVAILABILITY_IGNORED );
  proj_operation_factory_context_set_spatial_criterion( c, operationContext,  PROJ_SPATIAL_CRITERION_PARTIAL_INTERSECTION );
  PJ *src = proj_create_from_database( c, "EPSG", "3035", PJ_CATEGORY_CRS, false, nullptr );
  PJ *dest = proj_create_from_database( c, "EPSG", "5514", PJ_CATEGORY_CRS, false, nullptr );
  if ( PJ_OBJ_LIST *ops = proj_create_operations( c, src, dest, operationContext ) )
  {
    // all good!
  }
  else
  {
    // BAD!
    assert( false );
  }
```

I hit random behavior where sometimes this fails, and other times it correctly returns a handful of results.

From what I've been able to debug:

1. If I use a brand new context whenever I call this (`PJ_CONTEXT *c = proj_context_create();`) then there's never an issue, and the operations are always returned.

2. If I use an existing context, or a nullptr for the context, then sometimes the call fails with the error `internal_proj_create_operations: At least one of the operation lacks a source and/or target CRS`

It seems like some other operation performed using the context prior to this call is leaving it in an inconsistent state.

Although I can't 100% confirm it, I think this is caused by the following sequence of events:

1. Using a context, use proj_create_operations (as above) to get all operations between -s EPSG:3035 -t EPSG:5514.

2. iterate through these operations using proj_list_get. For each, call proj_as_proj_string. 

3. Some of the operations fail to export to proj, with the error: "Unimplemented" (maybe due to missing grids locally?)

4. Using the SAME context, try to repeat the call to proj_create_operations at some later stage -- the operation fails

This issue is the underlying cause of https://github.com/qgis/QGIS/issues/30569#issuecomment-557413847

### PR DESCRIPTION
**Pull request by rouault**
normalizeForVisualization(), promoteTo3D(), demoteTo2D(), alterGeodeticCRS(),
alterCSLinearUnit() and alterParametersLinearUnit() all used the object
returned by derivingConversionRef() to create a new ProjectedCRS. While doing
so, this caused the derivingConversion of the original object to have its
targetCRS set to the object returned by normalizeForVisualization() and similar.
If that object died, then the weak pointer would be reset, and the original
ProjectedCRS() has now its derivingConversionRef()->targetCRS() nullptr

So bottom line is use derivingConversion() for anything that is not pure
reading !!!

This is confirmed to be the fix for the QGIS scenario in
https://github.com/qgis/QGIS/issues/30569#issuecomment-540060919

In QGIS use case, the issue arised when using a projected CRS with a non-GIS
friendly axis (that is where normalizeForVisualization() created a new projectedCRS)

### CODE REVIEW

**Comment by nyalldawson** [comment]
Confirming that this fixes the issue for me too -- fantastic work @rouault !!

**Review by kbevers — APPROVED** [review]

**Comment by rouault** [comment]
Cherry-picked in 6.2 branch as bce4b158ab5f7d146de8e8fc98df4612dc8c2c9e

**Comment by rouault** [comment]
@kbevers   Jürgen Fischer announced on https://lists.osgeo.org/pipermail/qgis-developer/2019-November/059411.html that he had cherry-picked the fix to generate a 6.2.1 patched OSGeo4W package. So I guess we can probably just follow the original plan of a 6.3.0 release beginning of January. Speaking about that, PROJ master uses 7.0.0 numbering currently. Should it be changed to 6.3.0 or will you do so when you create the 6.3 branch from master ?

**Comment by kbevers** [comment]
@rouault Yes, I just saw Jürgens email. I agree with the 6.3.0 as the next release.

As for version number in master, yeah that should probably be changed. Feel free to do it yourself, otherwise I'll just do it when I prepare the release in about a months time.

**Comment by happydeb** [comment]
@nyalldawson @rouault @gioman So you all have made quite a bit of progress, I have reproduced some failures quite unintentionally since I hadn't followed the progress on this thread till now. I use Windows 10 64x. It started with adding multiple layers simultaneously...

Using ESRI 104019 (ITRF2014) (World - not Australia) GCRS as a Project Canvas CRS because I need to use projected data that will not work accurately with psuedo mercators (all tiled web maps Bing Google etc), and 3.10 is crashing with a projected canvas, i.e. with the Bing/Google Base map projected to match my data.

Adding GCRS layer worked fine on close and reopen.
Adding a different GRS layer from a data repsitory had a bug that I thought warranted closing QGIS to see if it would reopen. The problem happened when I added two layers simultaneously of different CGCRS (counties and roads) and only one showed up in the TOC but both rendered in the map Canvass (and in the overview insets in the layout)?! 

Reopening I chose the same GCRS 104019 world. QGIS crashed. Here is the error log.

Stack Trace


osgeo::proj::datum::Datum::conventionalRS :
osgeo::proj::datum::Datum::conventionalRS :
osgeo::proj::datum::Datum::conventionalRS :
OGRSpatialReference::~OGRSpatialReference :
OGRSpatialReference::operator= :
OGRGeomFieldDefn::~OGRGeomFieldDefn :
SHPWriteObject :
OGRFeatureDefn::~OGRFeatureDefn :
OGRFeatureDefn::~OGRFeatureDefn :
SHPWriteObject :
DBFWriteTuple :
OGREditableLayer::Translate :
QgsOgrProviderUtils::GDALCloseWrapper qgsogrprovider.cpp:4129
QgsConnectionPoolGroup::onConnectionExpired qgsconnectionpool.h:227
QMetaObject::activate :
QTimer::timerEvent :
QObject::event :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify qgsapplication.cpp:418
QCoreApplication::notifyInternal2 :
QEventDispatcherWin32Private::sendTimerEvent :
QEventDispatcherWin32::processEvents :
UserCallWinProcCheckWow :
DispatchMessageWorker :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main main.cpp:1598
BaseThreadInitThunk :
RtlUserThreadStart :




QGIS Info
QGIS Version: 3.10.0-A Coru�a
QGIS code revision: 6ffa89eb3e
Compiled against Qt: 5.11.2
Running against Qt: 5.11.2
Compiled against GDAL: 3.0.2
Running against GDAL: 3.0.2



System Info
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.18362

AND THE FOLLOWING HAPPENED BEFORE THE PROJECT REFERENCED ABOVE CRASHED QGIS. I DON'T KNOW IF ITS A SEPARATE OR RELATED ISSUE. TRYING TO RELOAD A SAVED LAYOUT 3.11 CANT FIND THEM, WHEN I MANUALLY INPUT THE LOCATION IT LOADS QGIS 3.46 (THE VERSION THE LAYOUT WAS MADE IN) THE LESSON IS DON'T TRY TO TRANSPORT ANY PART OF AN OLD PROJECT TO 3.1?

THE SPECIFIED FOLDER NOW SHOWS UP IN THE DEFAULT LAYOUT MGR. A CLOSE AND REOPEN AFTER THIS ISSUE WORKED (CHOOSING THE ESRI104019 (ITRF2014) BUT CRASHED ON THE MULTIPLE LAYERS ISSUE.

An error has occurred while executing Python code: 

RuntimeError: wrapped C/C++ object of type ContourGeneratorAlgorithm has been deleted 
Traceback (most recent call last):
  File "C:/Users/happy/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\contour\ContourGeneratorProvider.py", line 55, in loadAlgorithms
    self.addAlgorithm( alg )
RuntimeError: wrapped C/C++ object of type ContourGeneratorAlgorithm has been deleted


Python version: 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)] 
QGIS version: 3.10.0-A Coruña A Coruña, 6ffa89eb3e 

Python Path:
C:/Users/happy/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\contour
C:/PROGRA~1/QGIS3~1.10/apps/qgis/./python
C:/Users/happy/AppData/Roaming/QGIS/QGIS3\profiles\default/python
C:/Users/happy/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins
C:/PROGRA~1/QGIS3~1.10/apps/qgis/./python/plugins
C:\Program Files\QGIS 3.10\bin\python37.zip
C:\PROGRA~1\QGIS3~1.10\apps\Python37\DLLs
C:\PROGRA~1\QGIS3~1.10\apps\Python37\lib
C:\Program Files\QGIS 3.10\bin
C:\PROGRA~1\QGIS3~1.10\apps\Python37
C:\PROGRA~1\QGIS3~1.10\apps\Python37\lib\site-packages
C:\PROGRA~1\QGIS3~1.10\apps\Python37\lib\site-packages\win32
C:\PROGRA~1\QGIS3~1.10\apps\Python37\lib\site-packages\win32\lib
C:\PROGRA~1\QGIS3~1.10\apps\Python37\lib\site-packages\Pythonwin
C:/Users/happy/AppData/Roaming/QGIS/QGIS3\profiles\default/python
C:/Users/happy/Documents/KEC/20191009DeLuzWD/QGIS

CAN I IMPLEMENT ANY OF YOUR FIXES; WOULD IT HELP? OR SHOULD I JUST DO WORK AROUNDS?

**Comment by nyalldawson** [comment]
@happydeb just wait till 3.10.1 is released at the end of the week -- it will include this fix.

> RuntimeError: wrapped C/C++ object of type ContourGeneratorAlgorithm has been deleted

This is a bug in the "contour" plugin, you'll need to report this one with the authors of that plugin.
